### ISSUE TITLE
BUG: ezQuake silently crashes to desktop when a "bad" MDL file is loaded

### ISSUE THREAD

**Issue by Pulseczar1** [ezQuake/ezquake-source#412]
**"Please search for existing issues and check for potential duplicates before filing yours."**
I saw a very similar one about PNG files, but nothing about models.

**ezQuake version:
(Specify the Git commit hash if using a non-stable build - this is available through the `/version` command)**
I have duplicated the problem on ezQuake version 3.2.2, running on Windows 10.

> commit 37a3611697f9307f0d092759c171f9b71b3d2da6 (HEAD -> master, origin/master, origin/HEAD)
> Author: meag <meag@acm.org>
> Date:   Mon Oct 26 18:39:16 2020 +0000
>     FLASHBLEND: Fix when non-square sprite dimensions

I cloned the source code and compiled it with Visual Studio in order to identify the source of the crash. The crash also occurs with the Stable 3.2.2 and 3.1, available at http://www.ezquake.com/. I'm presuming the crash is due to the same thing in all three versions.

**OS/graphics card info:**
Let me know if more information is needed.

**Bug Description**
ezQuake crashes to the desktop without providing any error messages. A segmentation/read fault occurs. It appears to happen while loading game assets when connecting to a server using "bad" models. The "bad" model _can_ exist on the server, but only _has to_ exist on the client (or be downloaded from the server to the client), in order for the crash to occur. It probably also happens at map change, if an offending model wasn't used on the previous map, but is on the new map.

**Expected behavior / Proposed solution**
At minimum, gracefully deal with discrepancies in asset files, telling the user which files are problematic, if discrepancies are irreconcilable. Right now, users just crash without much of any clue as to why (at least in this case with this file).

**To Reproduce**
1. Put the attached **coil_BAD_FILE.mdl** in \<QuakeDir\>/fortress/progs/. Files: [bad_and_good_coil_mdl.zip](https://github.com/ezQuake/ezquake-source/files/5467645/bad_and_good_coil_mdl.zip)
2. Rename the file to **coil.mdl** inside fortress/progs/.
3. Connect to a Team Fortress server that loads a file by the name, **coil.mdl**. Custom Team Fortress (CustomTF) servers use a "coil.mdl" as a tesla coil. Connect to my server for an example server:  35.185.44.174:27500 For a server with fast downloads, instead, use:  gamehost2.tastyspleen.net:27507 (Both are CustomTF servers.) All files should appear to download as needed. You should not download a coil.mdl, since you have placed the bad one in your fortress/progs directory.
4. ezQuake crashes before "entering" the game (leaving the console). No error messages are seen.
5. Quarantine or delete the bad model file, to keep it from ever being duplicated and circulated.

**Screenshots, Configs, etc.**
Let me know if more files are required to duplicate the problem.

**Additional context**
I have not tested this on any other OS yet, nor any other version of ezQuake, nor any other client. I suspect this bug goes back a good ways with ezQuake and its two popular variants, nQuake and FortressOne, because this seems to be the bug that has been plaguing the Team Fortress community, especially FortressOne and the CustomTF For The People server, for a while. I suspect there are other "bad" models in the Team Fortress community (probably the entire Quake community) that also result in this crash-to-desktop. But this is the first model that has been identified, that I'm aware of, thanks to being able to finally duplicate the problem. I attached a working version of **coil.mdl**, for comparison.

Call stack at crash:
```
 	ezQuake_DEBUG.exe!GL_PrepareAliasModel(model_s * m, aliashdr_s * hdr) Line 112	C
 	ezQuake_DEBUG.exe!GLC_PrepareAliasModel(model_s * m, aliashdr_s * hdr) Line 407	C
 	ezQuake_DEBUG.exe!Mod_LoadAliasModel(model_s * mod, void * buffer, int filesize, const char * loadname) Line 865	C
 	ezQuake_DEBUG.exe!Mod_LoadModel(model_s * mod, qbool crash) Line 288	C
 	ezQuake_DEBUG.exe!Mod_ForName(const char * name, qbool crash) Line 312	C
 	ezQuake_DEBUG.exe!Model_NextDownload() Line 738	C
 	ezQuake_DEBUG.exe!Sound_NextDownload() Line 789	C
 	ezQuake_DEBUG.exe!CL_ParseModellist(qbool extended) Line 1795	C
 	ezQuake_DEBUG.exe!CL_ParseServerMessage() Line 3897	C
 	ezQuake_DEBUG.exe!CL_ReadPackets() Line 1573	C
 	ezQuake_DEBUG.exe!CL_Frame(double time) Line 2429	C
 	ezQuake_DEBUG.exe!Host_Frame(double time) Line 468	C
 	ezQuake_DEBUG.exe!WinMain(HINSTANCE__ * hInstance, HINSTANCE__ * hPrevInstance, char * lpCmdLine, int nCmdShow) Line 1307	C
 	[External Code]	
 	kernel32.dll![Frames below may be incorrect and/or missing, no symbols loaded for kernel32.dll]	Unknown
```

**Comment by jite** [comment]
Excellent bug report. I think @meag will be able to fix this quickly. Looks like some added sanity checks are in order :)

**Comment by Pulseczar1** [comment]
Thanks, jite. I'm usually on the receiving end of bug reports. That probably helps. :)

**Comment by meag** [comment]
Agree with @jite - thankyou for excellent bug report, really easy to reproduce.

"Quarantine or delete the bad model file, to keep it from ever being duplicated and circulated." - really like this idea, but it's a little tricky... at the moment, the client is simply detecting what files exist, downloads any missing, then goes through and tries to load them properly.  We can definitely improve that, but we'd also have to think about what to do with files which are in .pak/.pk3 etc.

**Comment by Pulseczar1** [comment]
Thank you, @meag , for such quick turnaround on this. I tried to test your code change. I just made your changes to the version of code that I have:  commit 37a3611697f9307f0d092759c171f9b71b3d2da6. But that version doesn't display any console text for me. So, I can't see the error message to see that your fix works for me. I can see that the game doesn't crash, though, when encountering the bad model, and instead disconnects from the server. I just can't see the error message. There are also a lot of missing things in the Escape menus as well. I have tried several different VS Solution Configurations and Platforms, to no avail. I am wondering if this commit just has an issue with it, and I should try a different version of the code.

On Step 5, where I said to quarantine or delete the bad model file, I was just talking about the error reproduction steps -- just a reminder to get rid of the file once you are done testing, in case it ever were to inadvertently get shared with others, if it were left in place.

Thanks again for being so quick to respond on this.

**Comment by meag** [comment]
Hi @Pulseczar1 - I don't have a good idea of why that is an issue, but I interested in finding out. Is this a problem you have with 3.6 in general?  If we could find what's causing it, that would be great.

Does it happen with a very simple quake directory?  (just the bare minimum in id1?)
What OS/gfx-card are you using?
Does it happen in the specific renderer builds?  (https://github.com/ezQuake/ezquake-source/releases/tag/latest)

Thanks,
meag

**Comment by Pulseczar1** [comment]
> Does it happen with a very simple quake directory? (just the bare minimum in id1?)
It does. I made a new directory and put one of the binaries I compiled, that has the problem, inside that directory. I then put an id1 directory inside the new directory, and inside the id1, I put pak0.pak and pak1.pak. Still no text in console and menus. All large text is in the menus, but small, console-like text is missing from menus.

I'll continue with your other questions.

**Comment by Pulseczar1** [comment]
OS is Windows 10. The graphics device on this machine is Intel HD Graphics 4600.

**Comment by meag** [comment]
If you do `/vid_restart`, do the graphics come back?
https://github.com/ezQuake/ezquake-source/issues/99
(May be a completely different issue)

**Comment by meag** [comment]
Sorry github garbled that, can you see if it's the same issue reported at https://github.com/ezQuake/ezquake-source/issues/99?

**Comment by Pulseczar1** [comment]
> Does it happen in the specific renderer builds?

What do you mean by that, exactly? I have tried "classic", "modern", and "all" Solution Configurations in Visual Studio, as well as 32 and 64 bit.

``/vid_restart`` appears to restart the game and then goes back to a blank console. (The console does have an image background.)

**Comment by Pulseczar1** [comment]
> Is this a problem you have with 3.6 in general? If we could find what's causing it, that would be great.

At the bottom of [this page](https://github.com/ezQuake/ezquake-source/releases/tag/latest), I downloaded ezquake-latest-win32.zip and ezquake-latest-windows-vstudio.zip. I tested each executable, within both zips (a total of 5 executables). I'm assuming these are the 3.6 version, but can't easily tell without being able to see output of ``/version``. They all have the problem.

3.2.2 and 3.1, downloaded from the official ezQuake website, does not have the problem, and they are running out of the same folder.

Things that might hint to the problem, that coincide with the problem:
- I have no mouse cursor while inside the Escape menus.
- There is no peculiar red screen that displays momentarily (flashes) normally when starting ezQuake versions that work. I have noticed this red screen for years, I think, but it does not show up in the "broken" executables.

If I go to a new game with a non-working executable, I enter map 'start'. I appeared to be spectator the first time, where I could fly around the level through walls, and had no gun view-model. Entering ``noclip`` and ``fly`` in the text-less console appears to have no bearing on this. Entered ``spectator 0`` in the console and the game locked up. But did respond to F10 to close??? Restarted the game, and I now have a gun and can't fly through walls, but all lights appear to be off (dark walls). Only things that are lit up appear to be things unaffected by lights (full-bright textures, like the sky and torches?). After I move and/or change my view orientation a good bit, I can trigger lights to turn on. I continue into e1m1. All sprites that show up with shotgun fire are strange black triangles, instead. Occasionally textures in the level black out, like floors.
